Add a runtime warning when allocating a float array.

nchataing opened the pull request:
Prints the call stack that lead to the allocation. It can be useful when wanting to migrate from the `float array` type to the `floatarray` one (and opting `--disable-flat-float-array`).

Here is an example :
```ocaml
let _ = Sys.enable_runtime_warnings true

let _ =
  let a = Array.make 1 0. in
  Printf.printf "%f\n" a.(0)
```
and the message :
```
Warning : creating a boxed float array
Raised by primitive operation at Test in file "test.ml", line 4, characters 10-25
0.000000
```

Commit:
Generate a runtime warning when allocating a boxed float array

nojb commented:
To give some further context: @nchataing recently completed an internship at @LexiFi where he worked on migrating our codebase to use `floatarray` in place of `float array` so that we could switch to `no-flat-float-array` mode.

As part of that work, he implemented this runtime warning which can be useful to spot allocations of `float array`s. These are not always easy to find (you may have fully polymorphic definitions which are instantiated at `float` far from the definition site).

For those interested, he also wrote a semi-automatic `.cmt`-based tool to assist the migration: https://github.com/nchataing/caml-migrate-floatarray

nojb commented on runtime/backtrace.c:
```diff
@@ -374,3 +374,38 @@ CAMLprim value caml_get_current_callstack(value max_frames_value)
 }
+
+/* Print a callstack backtrace */
+void caml_print_raw_backtrace(value bt)
```
If memory serves, this code was duplicated for simplicity (at the time we couldn't figure out how to reuse existing code). Suggestions how to do this would be welcome...

gasche replied:
This new code is a strict subset of what `caml_print_exception_backtrace` does, so you should be able to remove the duplication by calling the new function from `caml_print_exception_backtrace`?

yallop commented:
I'm not sure that this matters much, but this approach will also warn about values that can't be converted to `floatarray`, such as this one:

```ocaml
Array.make 1 (lazy 0.)
```

gasche commented:
Here is a slightly less devilish version of Jeremy's example:

```ocaml
type time = Time of float [@@unboxed]

let clock t = Array.make 1 (Time t)
```

I agree with him that this is not a big deal: the warning is only a heuristic (this could be clarified in the message), and we cannot detect those cases anyway (at type-checking time we may see an abstract variable, and at runtime the difference has been erased). It is arguably still useful to warn in those cases, as their performances also get degraded by the switch to `--no-flat-float-array`.

nojb commented:
> I'm not sure that this matters much, but this approach will also warn about values that can't be converted to `floatarray`, such as this one:
> 
> ```ocaml
> Array.make 1 (lazy 0.)
> ```

My understanding is that this is not possible if the compiler has been built in the `flat-float-array` mode (the current default), as in this case, the GC does *not* short-circuit values of type `float Lazy.t`.

yallop commented:
> >    Array.make 1 (lazy 0.)

> My understanding is that this is not possible if the compiler has been built in the `flat-float-array mode` (the current default), as in this case, the GC does _not_ short-circuit values of type `float Lazy.t`.

What I mean is:
- the code creates a boxed array in both `flat-float-array` and `no-flat-float-array` mode
- with this PR the compiler will warn that the code is creating a boxed float array, even though:
  * the code isn't creating a float array
  * there's nothing the user can do about it

alainfrisch commented:
The feature has proved to be very useful for our transition.  In terms of integrating it upstream, I think it could be worth dissociating the control flag from the usual runtime warnings (only about unclosed leaked file descriptors, IIUC), because, as discussed above, the "warning" is not always problematic and cannot be avoid in all situations.  In particular, it could be useful to manually silence the message in some (dynamic) scopes, where one knows that it can trigger.

gasche commented on runtime/backtrace.c:
```diff
@@ -100,28 +100,35 @@ static void print_location(struct caml_loc_info * li, int index)
-/* Print a backtrace */
-CAMLexport void caml_print_exception_backtrace(void)
+/* Print a callstack backtrace */
+void caml_print_raw_backtrace(int len, backtrace_slot * buffer)
```
I apologize for the nitpick, but I don't think the name of this function fits with its interface. I would expect `caml_print_raw_backtrace` to print a `raw_backtrace` value. I would suggest renaming it, for example `caml_print_backtrace_buffer`, and also passing the `int len` *after* the array, as I believe is more idiomatic in C.

(Otherwise the new version is fine, it nicely avoids any code duplication.)

nchataing replied:
Done, thanks for the review.

nojb commented on runtime/backtrace.c:
```diff
@@ -374,3 +381,20 @@ CAMLprim value caml_get_current_callstack(value max_frames_value)
+  bt = caml_get_current_callstack(Val_long(max_frames_value));
+
+  callstack_len = Wosize_val(bt);
+  bt_buffer = caml_stat_alloc(callstack_len);
```
Unless I missed something, there is no "free" corresponding to this "alloc".

nojb commented on runtime/backtrace.c:
```diff
@@ -374,3 +381,22 @@ CAMLprim value caml_get_current_callstack(value max_frames_value)
+  bt = caml_get_current_callstack(Val_long(max_frames_value));
+
+  callstack_len = Wosize_val(bt);
+  bt_buffer = caml_stat_alloc(callstack_len);
```
```suggestion
  bt_buffer = caml_stat_alloc(callstack_len * sizeof(backtrace_slot));
```
?

nchataing commented:
> The feature has proved to be very useful for our transition. In terms of integrating it upstream, I think it could be worth dissociating the control flag from the usual runtime warnings (only about unclosed leaked file descriptors, IIUC), because, as discussed above, the "warning" is not always problematic and cannot be avoid in all situations. In particular, it could be useful to manually silence the message in some (dynamic) scopes, where one knows that it can trigger.

As Alain mentioned, should I invest in disassociating this warning from the other runtime warnings ? I propose having similar to `Sys.enable_runtime_warnings` (another solution would be to use the `OCAMLRUNPARAM` if I'm not mistaken).

nojb commented on runtime/backtrace.c:
```diff
@@ -100,28 +100,35 @@ static void print_location(struct caml_loc_info * li, int index)
   if (!caml_debug_info_available()) {
-    fprintf(stderr, "(Cannot print stack backtrace: "
-                    "no debug information available)\n");
+    fprintf(stderr,
+            "(Cannot print stack backtrace: "
+            "no debug information available)\n");
```
No need for this diff.

nojb commented on runtime/backtrace.c:
```diff
@@ -374,3 +381,22 @@ CAMLprim value caml_get_current_callstack(value max_frames_value)
   CAMLreturn(res);
 }
+
+void caml_print_current_callstack(int max_frames_value)
```
Nitpicking, but I would call `max_frames_value` something like `num_frames` since it is not actually a "value".

Commit:
Dissociate the runtime warning specific to boxed float array allocations

from the other runtime warnings.

Also enable this warning only when in --disable-flat-float-array mode.

nchataing commented:
I've updated this PR by dissociating the wanted warning from the other warnings (new name `boxed_float_array_warning`). It can be enabled with similar functions from `Sys`, or with the `-F` option.

Also I've enabled this warning only when `--disable-flat-float-array` is set. Otherwise it could warn allocation of flat float arrays, which is (maybe?) not the wanted behavior (up to discussion).

nchataing commented:
@nojb can you take a second look into the code?

nojb commented on runtime/array.c:
```diff
@@ -249,6 +250,15 @@ CAMLprim value caml_make_vect(value len, value init)
       for (i = 0; i < size; i++) Field(res, i) = init;
     }
   }
+#ifndef FLAT_FLOAT_ARRAY
```
Something that worries me is that this warning is useful when working to migrate a codebase to `floatarray`. But it can be the case that for such a codebase, it is *not* possible to disable the `flat-float-array` mode before the migration is done. In that case, they would not be able to use the warning, right?

The only problem I see with allowing the warning even if the `flat-float-array` mode is enabled is that the message needs to be changed since with the mode enabled the allocated array is no longer "boxed".

nchataing replied:
Yes that's the problem that bothered me. Maybe we can just have a different message for each mode ?

nojb replied:
Yes, either that or we can just drop the word "boxed" from the message... But we should also think about the names of the API functions ...

nojb commented on stdlib/sys.mli:
```diff
@@ -354,6 +354,13 @@ val runtime_warnings_enabled: unit -> bool
+    boxed float arrays are allocated. *)
+
+val boxed_float_array_warning_enabled: unit -> bool
+(** Return whether the boxed float array warning is currently enabled. *)
```
"since" annotations missing.

nchataing replied:
What version should I put for this annotation? 4.13 is already in preparation so this will probably be merged only in 4.14, right?

nojb replied:
Yes, 4.14 sounds OK.

nojb commented on runtime/array.c:
```diff
@@ -249,6 +250,15 @@ CAMLprim value caml_make_vect(value len, value init)
+      && Is_block(init)
+      && Is_in_value_area(init)
+      && Tag_val(init) == Double_tag) {
+    fprintf(stderr, "Warning : creating a boxed float array\n");
```
The house style for the messages seems to be `"[ocaml] creating a boxed float array\n"`.

nojb commented on runtime/array.c:
```diff
@@ -249,6 +250,15 @@ CAMLprim value caml_make_vect(value len, value init)
+      && Is_in_value_area(init)
+      && Tag_val(init) == Double_tag) {
+    fprintf(stderr, "Warning : creating a boxed float array\n");
+    caml_print_current_callstack(10);
```
`10` here seems pretty arbitrary. One possibility would be to allow customizing the length of the back trace through the API.

Commit:
Add possibility to change the callstack size for the float array

allocation warning.

Rename "boxed_float_array" to "float_array_alloc".

nchataing commented:
I've renamed the warning from `boxed_float_array` to `float_array_alloc`, I think it makes more sense.

Also I've added a `set_float_array_alloc_callstack_size`. This makes for a long name however :-(

nojb commented on stdlib/sys.mli:
```diff
@@ -354,6 +354,22 @@ val runtime_warnings_enabled: unit -> bool
+    @since 4.14.0 *)
+
+val set_float_array_alloc_callstack_size: int -> unit
+(** Set the size of the printed backtrace when allocating a float array.
```
Instead of having a different call for this, we could just add an argument to `enable_float_array_alloc_warning`.

nojb commented on stdlib/sys.mli:
```diff
@@ -354,6 +354,22 @@ val runtime_warnings_enabled: unit -> bool
 
+val enable_float_array_alloc_warning: bool -> unit
+(** Control whether the OCaml runtime system can emit warnings when float
+    arrays are allocated.
```
Maybe it is better to be explicit and write `values of type [float array]` instead of `float arrays` to differentiate them from `floatarray` values.

nchataing replied:
Maybe we should be cautious about this. The example Gabriel provided emits the warning even though it is "not" a `float array` on the surface.

Oh well, I guess the main concern is to distinguish from `floatarray`

Commit:
Add callstack size argument to the enable function

Commit:
Add Changes entry

nojb commented on runtime/misc.c:
```diff
@@ -202,3 +202,18 @@ int caml_runtime_warnings_active(void)
+
+uintnat caml_float_array_alloc_warning = 0;
+static int caml_float_array_alloc_warning_first = 1;
+uintnat caml_float_array_alloc_callstack_size = 10;
```
You can merge `caml_float_array_alloc_warning` and `caml_float_array_alloc_callstack_size` in a single variable, where a value of `0` means "disabled" and anything "> 0" means enabled. That way you do not have to give a default value to `caml_float_array_alloc_callstack_size` and also you get to use the value passed in `OCAMLRUNPARAM`.

nchataing replied:
What should be the API for `Sys` in this case ? Just one argument with the "size" of the callstack ?

nojb replied:
I think the `Sys` API is fine as it is.

nchataing replied:
I mean we have to merge two arguments into one then.

nojb replied:
Merging both variables is more of an implementation detail. Semantically the data of whether the warning is enabled/disabled and the length of the backtrace are different. So the OCaml API can stay as it is, where the pair of arguments `(true, n)` maps to `n` on the C side and `(false, n)` maps to `0` on the C side. For extra safety the argument on the OCaml side may be normalized so that it is always `>= 1`.

Commit:
Merge callstack size variable into the float_array_alloc warning variable

nojb commented on Changes:
```diff
@@ -4,6 +4,9 @@ Working version
 
 ### Runtime system:
 
+- #10236 Add a runtime warning when allocating a float array.
```
I would add that the warning can be controlled by `Sys.enable_float_array_alloc_warning` or `OCAMLRUNPARAM=F=n`, as it may not be evident where to look for this warning otherwise.

nchataing replied:
I think the parameter is currently bugged, let me fix this.

nojb reviewed (commented):
It looks OK to me; the new `OCAMLRUNPARAM` setting needs to be added to the manual.

Since this touches the `stdlib` it may be good to get a second pair of eyes on it: @gasche or perhaps @alainfrisch ?

alainfrisch commented on stdlib/sys.mli:
```diff
@@ -354,6 +354,18 @@ val runtime_warnings_enabled: unit -> bool
+(** Control whether the OCaml runtime system can emit warnings when values
+    of type [float array] arrays are allocated. The second argument specifies
+    the size of the printed function callstack.
+
```
Is the second argument just ignored when the first argument is `false`?  This suggests changing the function to take an `int option` argument, no?  And, to be symmetric and allow overriding the setting for a limited dynamic scope, the getter function would also return a value of the same type.

alainfrisch replied:
(Or just use an `int` as on the C side, and document that <= 0 means the warning is disabled.)

nojb replied:
I argued against this, as it seemed cleaner to separate the two arguments, but sure, why not...

nchataing replied:
Just curiously, are there any option to deconstruct an option value in the C code ? (otherwise just a small wrapper around the C call suffices of course)

alainfrisch replied:
> I argued against this, as it seemed cleaner to separate the two arguments, but sure, why not...

The way I see it is that the two arguments are really related: one is useless for half the possible values of the other one :-)

If we think about the future addition of other similar warnings, with their own enabling/disabling functions, it could make sense to have just one global setting for the printed stack depth (used for all warnings).

nojb replied:
> > I argued against this, as it seemed cleaner to separate the two arguments, but sure, why not...
> 
> The way I see it is that the two arguments are really related: one is useless for half the possible values of the other one :-)

Yes, I agree that going with an option type is better.

nojb replied:
> Just curiously, are there any option to deconstruct an option value in the C code ? (otherwise just a small wrapper around the C call suffices of course)

Yes, you check if you are in the `None` case with `Is_block` and if you are in the `Some` case you can extract the payload with `Field(v, 0)`.

nchataing replied:
Nice, thanks for the tip. In terms of readability though, is it better to do this in the runtime part or in the stdlib part ?

nojb replied:
No strong opinion either way. Note that if you switch to an option, you will need to handle the case `Some n` when `n <= 0` and raise `invalid_arg` in that case.

gasche replied:
Note: we now have option-handling macros in mlvalues.h:

```
#define Val_none Val_int(0)
#define Some_val(v) Field(v, 0)
#define Tag_some 0
#define Is_none(v) ((v) == Val_none)
#define Is_some(v) Is_block(v)
```

(I think they were implemented by @nojb?)

nojb replied:
Indeed, ouch! :)

Commit:
Change the `Sys` to take an int option as argument.

Fix the control via OCAMLRUNPARAM.

Commit:
Update Changes

Commit:
Update stdlib .depend

nchataing commented:
The problem with this current semantic is that `None` and `Some 0` both maps to the same behavior, while we may want to have `Some 0` printing just a warning with no backtrace.

nojb commented:
> The problem with this current semantic is that `None` and `Some 0` both maps to the same behavior, while we may want to have `Some 0` printing just a warning with no backtrace.

How useful would this be? Supposedly one would always want to have some idea of where the allocation ocurred, in which case a stack trace of length 1 would be almost equivalent.

What I had in mind was to either forbid `n < 1` as arguments to `Some` or mapping it so that always `n >= 1` by eg replacing `n` by `max n 1`. It would be strange to have some values of `n` for which `Some n` and `None` are equivalent; if we wanted that, then it is better to use a simple `int` argument, the same as in the C side (as Alain suggested).

gasche commented on runtime/gc_ctrl.c:
```diff
@@ -787,3 +788,15 @@ CAMLprim value caml_ml_runtime_warnings_enabled(value unit)
+CAMLprim value caml_ml_float_array_alloc_warning_enabled(value unit)
+{
+  CAMLassert (unit == Val_unit);
+  return Val_bool(caml_float_array_alloc_warning);
```
This seems inconsistent with the use of `Int_val ` above.

nchataing replied:
Fixed, thanks!

gasche commented:
Two competing interface proposals is maybe not enough, let's add a third one:
- no call stacks shown by default
- show call stacks of maximal length `max_int` when `OCAMLRUNPARAM=b` is active

What is the reason for trying to let users select shorter trace? UI convenience, or performance?

nojb commented:
> What is the reason for trying to let users select shorter trace? UI convenience, or performance?

Mostly UI convenience I would say.

Commit:
Fixes

Commit:
update depend

nchataing commented:
Last round of review @nojb @alainfrisch?

nojb reviewed (commented):
LGTM

@alainfrisch @gasche are you happy with the current API signatures?

gasche commented:
I don't understand why `Some 0` is forbidden, I would expect it to print a warning without a callstack.

alainfrisch commented on stdlib/sys.mli:
```diff
@@ -354,6 +354,19 @@ val runtime_warnings_enabled: unit -> bool
     @since 4.03.0 *)
 
+val enable_float_array_alloc_warning: int option -> unit
+(** Control whether the OCaml runtime system can emit warnings when values
```
```suggestion
(** Control whether the OCaml runtime system emits warnings when values
```

nchataing replied:
I was copying the above `enable_runtime_warnings` comment but this proposition looks better indeed.

alainfrisch commented on stdlib/sys.mli:
```diff
@@ -354,6 +354,19 @@ val runtime_warnings_enabled: unit -> bool
+val enable_float_array_alloc_warning: int option -> unit
+(** Control whether the OCaml runtime system can emit warnings when values
+    of type [float array] arrays are allocated.
+    None   -> the warning is disabled
```
I'm not sure how this ASCII art would render in the generated documentation.  It would be good to mention the constraint on [s], and the exception being generated when it isn't satisfied.

nchataing commented:
> I don't understand why `Some 0` is forbidden, I would expect it to print a warning without a callstack.

With @gasche we agreed on the fact that we should have this behavior. It is possible by changing a bit the C variable so that :
- `-1` means the warning is disabled;
- `>= 0` means the warning is enabled with the callstack size being the variable.

Commit:
Change acceptable threshold value to 0

Commit:
Tweak comments

nchataing committed:
Merge remote-tracking branch 'ocaml/trunk' into runtime_warnings

nchataing committed:
fix: remove double free & use correct formatter for intnat

nchataing committed:
chore: update Changes entry

nchataing committed:
fix: hygiene

nchataing commented:
Attempting to revive this PR. I am no longer sure of the utility of such a feature, but it gave me a good context to get back into a bit of OCaml hacking, so here goes nothing.

---

With the arrival of version 5 of the compiler the runtime for backtraces changed a bit so I had to modify a bit the original implementation.
The latest proposed interface is implemented : 
```ocaml
# Sys.enable_float_array_alloc_warning;;
- : int option -> unit = <fun>
# Sys.float_array_alloc_warning_enabled;;
- : unit -> int option = <fun>
```
Alternatively, one can provide `OCAMLRUNPARAMS=F=x` to display the warning with a bactrace of `x` elements.
Note that this backtrace will be displayed **only if** `b=1` is also provided to the `ocamlrun` params.

cc @nojb @gasche

nojb commented on Changes:
```diff
@@ -3191,6 +3198,7 @@ OCaml 5.0.0 (15 December 2022)
   bytecode interpreter.
   (David Allsopp and Xavier Leroy, report by Samuel Hym, review by Xavier Leroy
   and Gabriel Scherer)
+>>>>>>> ocaml/trunk
```
Merge left-over.

nojb reviewed (commented):
Hello! I am off today, but I will try to review sometime in the next few days.

nchataing committed:
chore: remove merge left-over

nojb commented on stdlib/sys.mli:
```diff
@@ -417,6 +417,20 @@ val runtime_warnings_enabled: unit -> bool
+val float_array_alloc_warning_enabled: unit -> int option
+(** Return whether the float array allocation warning is currently enabled.
+
+    @since 5.2 *)
```
```suggestion
    @since 5.4 *)
```

nojb commented on runtime/startup_aux.c:
```diff
@@ -103,6 +103,9 @@ void caml_parse_ocamlrunparam(void)
       case 'c': scanmult (opt, &params.cleanup_on_exit); break;
       case 'd': scanmult (opt, &params.max_domains); break;
       case 'e': scanmult (opt, &params.runtime_events_log_wsize); break;
+      case 'F':
+        // scanned value is positive in case we declare the warning
+        scanmult (opt, (uintnat*)&caml_float_array_alloc_warning); break;
```
Nit, but I would remove the comment; it is not clear what it is referring to without more context (which warning?).
```suggestion
      case 'F': scanmult (opt, (uintnat*)&caml_float_array_alloc_warning); break;
```

nojb commented on runtime/caml/misc.h:
```diff
@@ -688,6 +688,9 @@ void caml_gc_log (const char *, ...)
 CAMLextern int caml_runtime_warnings_active(void);
 
+extern intnat caml_float_array_alloc_warning;
+int caml_float_array_alloc_warning_active(void);
```
I guess `CAMLextern` should be added here? (By analogy with `caml_runtime_warnings_active`).
```suggestion
CAMLextern int caml_float_array_alloc_warning_active(void);
```

nojb commented on runtime/backtrace.c:
```diff
@@ -100,8 +100,8 @@ static void print_location(const struct caml_loc_info * li, int index)
 
-/* Print a backtrace */
-CAMLexport void caml_print_exception_backtrace(void)
+/* Print a callstack backtrace */
```
Let's remove this misleading comment (nothing about callstacks here).
```suggestion
```

nojb commented on Changes:
```diff
@@ -110,6 +110,13 @@ Working version
 
+- #10236 Add a runtime warning when allocating a float array, which prints
+  a backtrace of the current callstack.
+  It can be controled through `Sys.enable_float_array_alloc`, or
```
```suggestion
  It can be controled through `Sys.enable_float_array_alloc_warning`, or
```

nojb commented on stdlib/sys.mli:
```diff
@@ -417,6 +417,20 @@ val runtime_warnings_enabled: unit -> bool
+    warning  is enabled with a backtrace of size [s].
+    @raise Invalid_argument if called with [Some s] and [s < 0]
+
+    @since 5.2 *)
```
```suggestion
    @since 5.4 *)
```

nojb commented on stdlib/sys.mli:
```diff
@@ -417,6 +417,20 @@ val runtime_warnings_enabled: unit -> bool
+
+    @since 5.2 *)
+
+val float_array_alloc_warning_enabled: unit -> int option
```
What would be the use of this function?

nojb commented on runtime/backtrace.c:
```diff
@@ -100,8 +100,8 @@ static void print_location(const struct caml_loc_info * li, int index)
-/* Print a backtrace */
-CAMLexport void caml_print_exception_backtrace(void)
+/* Print a callstack backtrace */
+void caml_print_backtrace_buffer(backtrace_slot * buffer, int len)
```
Something wrong's here: `buffer` and `len` are unused.

nojb reviewed (commented):
I took a look. Technically the PR looks roughly OK (modulo some nits below). But to be honest, I am not sure there is much use for this warning. Firstly because very few users (as far as a I know) actually switch to the "no-flat-float-array" mode (and what is the use of the warning otherwise?). @nchataing: do you remember if you used this warning when doing the migration at LexiFi? In any case, we have never used it since... That said, the patch is pretty small and basically just moves code around, so it is not expensive to merge.

In short, I am not convinced it is worth getting it in, and would appreciate a second opinion by another core dev before moving forward with it.

nchataing commented:
> do you remember if you used this warning when doing the migration at LexiFi?

As far as I remember, we used it as validation that we correctly migrated every possible instances of `float array` to `floatarray`, modulo generic usage of `'a array`.
So it was useful as a hint only, and probably the tool that we published in the end would be largely enough if anyone is interested in doing this kind of migration.
